### PR TITLE
WebGPU: perf optimization & misc

### PR DESCRIPTION
**Pull request by Popov72**
(sorry for the big PR!)

* added time query to WebGPU to have meaningful values in "GPU Frame Time" in the inspector
* added `engine.compatibilityMode = false` ("fast path") and `engine.snapshotRendering/engine.snapshotRenderingMode` as ways to improve perfs:
  * `engine.compatibilityMode = false`: will create a bundle the first time a draw is done and will be reused afterwards. Preliminary work
  * `engine.snapshotRendering = true` enables a mode where the GPU commands of the next frame are recorded in bundle lists and replayed in subsequent frames. When `engine.snapshotRenderingMode=1` we also don't execute some JS code, like the `_evaluateActiveMeshes` and the mesh `render` methods. This is meant to improve perf in mostly static scenes (like the forest demo, where fps > 60fps - GPU bound)
* did some small changes to improve perf in the legacy mode ("slow path")
* added `alphaCutOff` and `vAmbientColor` to the ubo for the standard material, so that all standard parameters of the material are in the ubo and not in the LeftOver ubo
* made some changes in shadow generator/CSM + `ShadowDepthWrapper` to be able to use one `DrawWrapper` per cascade in CSM mode (to enable fast path to work with CSM and `ShadowDepthWrapper`) and one `DrawWrapper` per cube face for point lights (in standard shadow generator)
* added a `checkUbosContentBeforeUpload` feature. When true, it checks that the current content of the ubo is really different from the content of the corresponding GPU buffer before uploading. We need this in WebGPU to avoid having a lot of unncessary uploads (that WebGL does not have because Angle is doing some smart job about this)
* fixed a bug where `ShadowGenerator.bindShadowLight` was called for a material even if the mesh this material was applied to was not a shadow receiver
* synced with latest WebGPU spec

### CODE REVIEW

File: src/Engines/Extensions/engine.occlusionQuery.ts
Comment (deltakosh) [inline src/Engines/Extensions/engine.occlusionQuery.ts:316]: does it work when using occlusion query? I was under the impression that only one query was available at time

File: src/Engines/constants.ts
Comment (deltakosh) [inline src/Engines/constants.ts:602]: will this be enabled when users call scene.frezzeActiveMeshes?

Comment (Popov72) [reply src/Engines/constants.ts:602]: No, it's independent from freezeActiveMeshes. It's activated when setting `engine.snapshotRendering = true`.

File: src/Engines/Extensions/engine.occlusionQuery.ts
Comment (Popov72) [reply src/Engines/Extensions/engine.occlusionQuery.ts:316]: Not occlusion query but timer query. But it was already there before :) In this file, there were occlusion + time query functions

File: src/Engines/constants.ts
Comment (deltakosh) [reply src/Engines/constants.ts:602]: do you think it is an option to do i? Does it make sense?

File: src/Engines/Extensions/engine.occlusionQuery.ts
Comment (deltakosh) [reply src/Engines/Extensions/engine.occlusionQuery.ts:316]: oh right..I'm a dick with file names

can you rename the file name to something simply like engine.query ?

File: src/Engines/constants.ts
Comment (Popov72) [reply src/Engines/constants.ts:602]: No, I think it's really two different things, you could want freezing active meshes without enabling snapshot rendering as snapshot rendering have more constraints than freezing meshes (meaning, when snapshot rendering is enabled a number of things won't work as expected if you do them that would work if you only froze active meshes).

**Comment by Popov72** [comment]
Compilation fails with:
image

I have added `es2020.bigint` in the ts lib, I guess it's what makes the compil fail but I have no idea why.

Maybe @sebavan you know?

**Comment by Popov72** [comment]
Everything is ok except for these errors in "Unit tests" and "Module tests":
```
/home/vsts/work/1/s/dist/preview release/babylon.module.d.ts(86999,18): error TS2300: Duplicate identifier 'Observer'.
/home/vsts/work/1/s/dist/preview release/babylon.module.d.ts(145969,10): error TS2300: Duplicate identifier 'Observer'.
```

**Comment by Popov72** [comment]
Ok, problems fixed, I switched the PR to "Ready for review"

**Comment by deltakosh** [comment]
Waiting for @sebavan approval

File: src/scene.ts
Comment (sebavan) [inline src/scene.ts:774]: Why static if we pass the scene in ?

File: materialsLibrary/src/cell/cellMaterial.ts
Comment (sebavan) [inline materialsLibrary/src/cell/cellMaterial.ts:266]: Why static if it is on scene ? I ll put the comment in scene as well

File: src/Engines/WebGPU/webgpuBundleList.ts
Comment (sebavan) [inline src/Engines/WebGPU/webgpuBundleList.ts:172]: Could we init the array at the right size ? (prevents over allocation when it grows)

File: src/Materials/material.ts
Comment (sebavan) [inline src/Materials/material.ts:995]: Same here for the static ?

File: src/scene.ts
Comment (sebavan) [inline src/scene.ts:800]: Same static comment but not sure why it had been moved to scene ?

**Review by sebavan — COMMENTED** [review]
Sounds goood aside the static move to Scene ?

File: src/scene.ts
Comment (Popov72) [reply src/scene.ts:774]: I preferred to retain the exact same signature to avoid too many breaking changes, but the move is a breaking change itself so you're right it's better to just make it non static. Going to perform the change.

File: src/scene.ts
Comment (Popov72) [reply src/scene.ts:800]: Because of circular dependency problems. I had to move it along with `BindEyePosition` as the latter is calling `FinalizeSceneUbo`.

File: materialsLibrary/src/cell/cellMaterial.ts
Comment (Popov72) [reply materialsLibrary/src/cell/cellMaterial.ts]: Not static anymore

File: src/Engines/WebGPU/webgpuBundleList.ts
Comment (Popov72) [reply src/Engines/WebGPU/webgpuBundleList.ts]: Done

File: src/Materials/material.ts
Comment (Popov72) [reply src/Materials/material.ts]: Not static anymore

**Comment by sebavan** [comment]
> Compilation fails with:
> image
> 
> I have added `es2020.bigint` in the ts lib, I guess it's what makes the compil fail but I have no idea why.
> 
> Maybe @sebavan you know?
> 
> [...] Trying to add "es2015.symbol.wellknown", let's see how it comes...

No Idea :-) Let s check offline

**Review by sebavan — APPROVED** [review]

**Comment by Popov72** [comment]
> No Idea :-) Let s check offline

That's ok, there's no error anymore.

**Comment by Popov72** [comment]
If ok, can this be merged?

**Pull request merged**
